egs/librispeech/ASR/pruned_transducer_stateless2/scaling.py

```diff
@@ -111,6 +111,74 @@ def backward(
         return x_grad - neg_delta_grad, None, None, None, None, None, None
 
 
+class GradientFilterFunction(torch.autograd.Function):
+    @staticmethod
+    def forward(
+        ctx,
+        x: Tensor,
+        batch_dim: int,  # e.g., 1
+        threshold: float,  # e.g., 10.0
+        *params: nn.Parameter,  # module parameters
+    ):
+        if x.requires_grad:
+            if batch_dim < 0:
+                batch_dim += x.ndim
+            ctx.batch_dim = batch_dim
+            ctx.threshold = threshold
+        return x, *params
+
+    @staticmethod
+    def backward(
+        ctx,
+        x_grad: Tensor,
+        *param_grads: nn.Parameter,
+    ):
+        dim = ctx.batch_dim
+        norm_dims = [d for d in range(x_grad.ndim) if d != dim]
+        norm_of_batch = (x_grad ** 2).mean(dim=norm_dims, keepdim=True).sqrt()
+        median_norm = norm_of_batch.median()
+        mask = norm_of_batch <= median_norm * ctx.threshold
+        x_grad = x_grad * mask
+
+        if not torch.all(mask):
```

Comment:
it would probably be more efficient, if we are using GPU to not have a conditional here, and to use multiplication, so we don't have to sync the GPU.
Could do:
```
   all = torch.all(mask).to(x_grad.dtype)  # 1 if no grad was zeroed, 0 if any was zeroed
   param_grads = [ all * x for x in param_grads ]
```
... OK to use this code if on CPU though.

Reply:
OK i see.